components/builder-core/src/access_token.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+use chrono::{self, Utc, Duration};
+
+use hab_net::privilege::FeatureFlags;
+use error::{Error, Result};
+use integrations::{encrypt, decrypt};
+use protocol::{message, sessionsrv};
+
+pub const BUILDER_ACCOUNT_ID: u64 = 0;
+const ACCESS_TOKEN_PREFIX: &'static str = "TOKN";
+const BUILDER_TOKEN_LIFETIME_HOURS: i64 = 2;
+
+pub fn generate_bldr_token(key_dir: &PathBuf) -> Result<String> {
+    generate_access_token(
+        key_dir,
+        BUILDER_ACCOUNT_ID,
+        FeatureFlags::all().bits(),
+        Duration::hours(BUILDER_TOKEN_LIFETIME_HOURS),
+    )
+}
+
+pub fn generate_user_token(key_dir: &PathBuf, account_id: u64, privileges: u32) -> Result<String> {
+    generate_access_token(
+        key_dir,
+        account_id,
+        privileges,
+        Duration::max_value(), // User tokens never expire, can only be revoked
+    )
+}
+
+pub fn generate_access_token(
+    key_dir: &PathBuf,
+    account_id: u64,
+    flags: u32,
+    lifetime: Duration,
+) -> Result<String> {
+    let expires = Utc::now()
+        .checked_add_signed(lifetime)
+        .unwrap_or(chrono::MAX_DATE.and_hms(0, 0, 0))
+        .timestamp();
+
+    let mut token = sessionsrv::AccessToken::new();
+    token.set_account_id(account_id);
+    token.set_flags(flags);
+    token.set_expires(expires);
+
+    let bytes = message::encode(&token).map_err(Error::Protocol)?;
+    let ciphertext = encrypt(key_dir, &bytes)?;
+
+    Ok(format!("{}{}", ACCESS_TOKEN_PREFIX, ciphertext))
+}
+
+pub fn is_access_token(token: &str) -> bool {
+    token.starts_with(ACCESS_TOKEN_PREFIX)
+}
+
+pub fn validate_access_token(key_dir: &PathBuf, token: &str) -> Result<sessionsrv::Session> {
+    assert!(is_access_token(token));
+
+    let encoded = token.to_owned().split_off(ACCESS_TOKEN_PREFIX.len());
+
+    let bytes = decrypt(key_dir, &encoded)?;
+
+    let payload: sessionsrv::AccessToken = match message::decode(&bytes) {
+        Ok(p) => p,
+        Err(e) => {
+            warn!("Unable to deserialize access token, err={:?}", e);
+            return Err(Error::TokenInvalid);
+        }
+    };
+
+    let now = Utc::now().timestamp();
+
+    if payload.get_expires() < now {
```

Comment:
This might seem like overkill, but I like taking maximum advantage of type-safety by comparing the `DateTime`s rather than raw integers:
```rust
    let expiration = Utc.timestamp(payload.get_expires(), 0 /* nanoseconds */);

    if expiration < Utc::now() {
```

Reply:
That seems reasonable.. the only caveat is that looks like Utc.timestamp(..) can panic.  In that case the direct compare would probably give undefined behavior.  However for the case of user tokens, the undefined behavior would be _more_ conservative, ie, we would not allow use of the token.  So on the whole, we're probably better off with the direct compare here I think.

Reply:
I don't understand why we can conclude that the undefined behavior would be more conservative. Let's consider the situations where `Utc.timestamp()` panics:
1. Out-of-range number of seconds
2. invalid nanosecond

#2 is a non-issue since we're using literal 0. #1 Should only be a problem if we end up outside the representable range that `chrono` supports (roughly 262145 BCE–262143 CE). If the expiration timestamp in the payload represented a time _before_ that, we would not allow use of the token, but if it represented a time _after_ we *would*. Now, we don't expect to ever be able to get those sorts of values short of a logic error or data corruption. In those cases, I think panicking may be appropriate. However, if we want to not panic, we can use `timestamp_opt()` to ensure we deny access when the unexpected occurs and allow us to make a note of the unexpected situation without crashing:
```rust
    match Utc.timestamp_opt(payload.get_expires(), 0 /* nanoseconds */) {
        Some(time) => {
            if time < Utc::now() {
                Err(Error::TokenExpired)
            } else {
                Ok(payload.into())
            }
        }
        _ => Err(Error::TokenInvalid),
    }
```

Reply:
👍 to timestamp_opt - that works.  Yeah, definitely want to avoid panics if at all possible.  Since user tokens don't expire, we don't have to worry about allowing access if the seconds are out of range - revocation won't depend on this.  Builder tokens do expire, so that's the most of the exposure.